### ISSUE TITLE
Car stops after I think roughly one lap of the track

### ISSUE THREAD

**Issue by stela** [stela/CarND-Capstone#7]
After driving apparently successfully around the simulator track, the car brakes. This is what is logged, notice the light_wp changing from 9733 to -1:
```
[WARN] [1531321775.286391]: image_cb: processed light_wp: 9733, state: 4
[WARN] [1531321775.874234]: Color detected: 4
[WARN] [1531321775.874739]: image_cb: processed light_wp: 9733, state: 4
[WARN] [1531321776.452423]: Color detected: 4
[WARN] [1531321776.452859]: image_cb: processed light_wp: 9733, state: 4
[WARN] [1531321777.089557]: Color detected: 0
[WARN] [1531321777.090030]: image_cb: processed light_wp: 9733, state: 0
[WARN] [1531321777.654192]: Color detected: 0
[WARN] [1531321777.654874]: image_cb: processed light_wp: 9733, state: 0
[WARN] [1531321778.295658]: Color detected: 2
[WARN] [1531321778.296362]: image_cb: processed light_wp: 9733, state: 2
[WARN] [1531321778.872563]: Color detected: 2
[WARN] [1531321778.875110]: image_cb: processed light_wp: 9733, state: 2
[WARN] [1531321779.510623]: Color detected: 2
[WARN] [1531321779.511406]: image_cb: processed light_wp: 9733, state: 2
[INFO] [1531321779.511741]: LiveDetect: True | TrafficLight : green
[WARN] [1531321780.069390]: Color detected: 2
[WARN] [1531321780.070332]: image_cb: processed light_wp: 9733, state: 2
[WARN] [1531321780.693575]: Color detected: 2
[WARN] [1531321780.693940]: image_cb: processed light_wp: 9733, state: 2
[WARN] [1531321781.286760]: Color detected: 2
[WARN] [1531321781.287090]: image_cb: processed light_wp: 9733, state: 2
[WARN] [1531321781.900078]: Color detected: 2
[WARN] [1531321781.900487]: image_cb: processed light_wp: 9733, state: 2
[WARN] [1531321782.485680]: Color detected: 2
[WARN] [1531321782.485993]: image_cb: processed light_wp: 9733, state: 2
[WARN] [1531321783.103988]: Color detected: 2
[WARN] [1531321783.104410]: image_cb: processed light_wp: 9733, state: 2
[WARN] [1531321783.699722]: Color detected: 2
[WARN] [1531321783.700057]: image_cb: processed light_wp: 9733, state: 2
[WARN] [1531321784.292047]: Color detected: 2
[WARN] [1531321784.292560]: image_cb: processed light_wp: 9733, state: 2
[WARN] [1531321784.897244]: Color detected: 2
[WARN] [1531321784.898027]: image_cb: processed light_wp: -1, state: 2
[WARN] [1531321785.484816]: Color detected: 4
[WARN] [1531321785.485418]: image_cb: processed light_wp: -1, state: 4
[WARN] [1531321786.106197]: Color detected: 4
[WARN] [1531321786.106740]: image_cb: processed light_wp: -1, state: 4
[WARN] [1531321786.689877]: Color detected: 4
[WARN] [1531321786.690312]: image_cb: processed light_wp: -1, state: 4
[INFO] [1531321786.690697]: LiveDetect: True | TrafficLight : unknown
[WARN] [1531321787.302523]: Color detected: 4
[WARN] [1531321787.307568]: image_cb: processed light_wp: -1, state: 4
[WARN] [1531321787.886949]: Color detected: 4
[WARN] [1531321787.887507]: image_cb: processed light_wp: -1, state: 4
[WARN] [1531321788.456066]: Color detected: 4
[WARN] [1531321788.456450]: image_cb: processed light_wp: -1, state: 4
[WARN] [1531321789.122437]: Color detected: 4
[WARN] [1531321789.123057]: image_cb: processed light_wp: -1, state: 4
[WARN] [1531321789.690687]: Color detected: 4
[WARN] [1531321789.691214]: image_cb: processed light_wp: -1, state: 4
[WARN] [1531321790.337479]: Color detected: 4
[WARN] [1531321790.338518]: image_cb: processed light_wp: -1, state: 4
[WARN] [1531321790.901133]: Color detected: 4
[WARN] [1531321790.901463]: image_cb: processed light_wp: -1, state: 4
[WARN] [1531321791.486448]: Color detected: 4
[WARN] [1531321791.486826]: image_cb: processed light_wp: -1, state: 4
```
I would assume it must have passed a traffic light, but at the place the car stopped there's just a "speed limit 50" board about 10 meters in front of it, see nothing behind it.

**Comment by stela** [comment]
Happened again when driving a lap, did see it driving with the green waypoint dots ahead of it for a while after it started logging light_wb=-1, and then the green waypoint-dots disappeared and the simulator says: Brake: 700, so it appears it thinks it should brake

**Comment by stela** [comment]
Seems it drove "off the world" and without waypoints braked. I switched to manual driving and just had to drive tens of meters to get a positive light_wp, and then I could switch to automatic driving and it would drive by itself. Think we need to treat the waypoints as a circular list or something, this is probably the problematic non-circular code:
https://github.com/stela/CarND-Capstone/blob/3f27217a0aee1213846502261f6cbba50fbe318f/ros/src/waypoint_updater/waypoint_updater.py#L87-L94
Should append waypoints from 0 and onwards to always return LOOKAHEAD_WPS number of points.

Second, in that code `self.stopline_wp_idx == -1` which shouldn't happen, right? The code "guilty" of that must be the subscriber callback, it's never assigned anywhere else after initialization:
https://github.com/stela/CarND-Capstone/blob/3f27217a0aee1213846502261f6cbba50fbe318f/ros/src/waypoint_updater/waypoint_updater.py#L126-L128

Or is it by design that the car stops after one lap? It kind of seems like that since the -1 checking code was from the walkthrough. It kicks in both after initialization before the first traffic_cb invocation happened as well as at the end of the track.

**Comment by StefanGerlach** [comment]
No its not by design, that the car stops there. We have to treat the course as a circular (like it is).

**Comment by stela** [comment]
Working on a fix for this on the multiple_laps branch. Ended up installing PyCharm in the Linux ROS VM in order to be able to create and run a unit test for this. Mounting a folder from my Mac host didn't work through virtualbox, as it messes up the permissions of files (e.g. the execute-flag must not be set on unit test .py files)

**Comment by stela** [comment]
Here is a discussion on slack regarding what do do at the end of the track, with varying opinions about how to handle it.
https://carnd.slack.com/archives/C6NVDVAQ3/p1505320782000298
It says speed is set to 0 at the last waypoint provided, and that causes the car to stop, even when wraparound is handled correctly. By overriding the speed one can make the car continue driving. Also noticed that just 20 waypoints lookahead seemed to be too little to brake safely sometimes in the simulator, so made it 40 instead.

It works for me now in the simulator, will create a pull request for this now.

**Comment by stela** [comment]
Resolved by #9, closing it.

**Comment by stela** [comment]
We should probably reverse this change if we need to re-submit, apparently the car _is_ supposed to respect the 0 waypoint speeds, see https://carnd.slack.com/archives/C6NVDVAQ3/p1522261806000786
>  Stephen Welch
> Hi All! The vehicle should follow the speed set for each waypoint (in this case coming to a stop at the end of the waypoints). We don't expect the vehicle to be able to continuously drive in a loop. However, we may try to restart the dbw in the middle of the waypoint set after driving the loop once, if needed. (e.g. if we want to get another shot at the traffic light.)

_From a thread in #s-t3-p-system-integra, Mar 28th at 20:30_
